### PR TITLE
Space - Leah

### PR DESCRIPTION
**Pull request by leahwho**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | The biggest challenge was trying to decide where to store reservations and ultimately realizing that my final decision was the wrong one!
What was a design decision you made that changed over time over the project? | Initially my design did not include a Room class.  I thought I could capture everything using just a FrontDesk, Reservation, & Date Class.  However I opted to create this class so I could more easily separate each rooms' reservations - I very much regret this decision as it made accessing those reservations quite complex.
What was a concept you gained clarity on, or a learning that you'd like to share? | I am solidifying my understanding of composition and really am coming to understand how the connections between classes are a very important part of your initial design.  I also saw the benefits of regularly reviewing the user stories - this helped me stay on track and not get in the weeds.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | I tested correctly instantiating all my classes - these are nominal tests because the input is all expected (e.g. room number is passed as a number, dates are passed as Dates).
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | I used several input validators, for example in the FrontDesk class, I wanted to make sure that the number of hotel rooms someone could instantiated would be positive.  I don't feel like I did a good enough job of writing tests for most edge cases.  Although I'm getting better at testing, I feel most confident with coming up with nominal tests.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I got a lot of practice with this process and it felt a lot easier after the first day of forcing myself to do it.  I gained some insight that if I'm not willing to pseudocode or write tests before writing code that that's a sign I need to take a break.  It's coming a lot more naturally to me with all this practice.

### CODE REVIEW

File: lib/date_range.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        raise ArgumentError, "End date cannot be before start date"
+      end
+    end
+    
+    
+    def validate_date_object
+      unless @check_in.class == Date && @check_out.class == Date
+        raise ArgumentError, "Please enter a valid date!"
+      end
+    end
```
Comment (kaidamasaki) [inline lib/date_range.rb:27]: Since these are very short and only called once I would probably just include them in the body of your `initialize` method.

File: lib/date_range.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      @check_out = check_out
+      
+      validate_date_range
+      validate_date_object
+    end
+    
+    
+    def validate_date_range
+      if @check_out < @check_in
+        raise ArgumentError, "End date cannot be before start date"
```
Comment (kaidamasaki) [inline lib/date_range.rb:18]: It's helpful to include the bad argument in the message when raising an `ArgumentError`:

```suggestion
        raise ArgumentError, "End date (#{end_date}) cannot be before start date (#{start_date})"
```

File: lib/date_range.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    def validate_date_object
+      unless @check_in.class == Date && @check_out.class == Date
+        raise ArgumentError, "Please enter a valid date!"
+      end
+    end    
+    
+
+    def num_nights
+      num_nights = (@check_out - @check_in).to_i
+      return num_nights
```
Comment (kaidamasaki) [inline lib/date_range.rb:32]: You don't need to assign to a variable before returning a value:

```suggestion
      return (@check_out - @check_in).to_i
```

File: lib/date_range.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    end  
+    
+    
+    # User story: I can access the list of reservations for a specific date, so that I can track reservations by date
```
Comment (kaidamasaki) [inline lib/date_range.rb:36]: I like that you included the user stories as comments!  Clever. 😃

File: lib/date_range.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    def date_in_range?(date)
+      range = @check_in...@check_out
+      
+      if range.include?(date)
+        return true # there is an existing reservation for tonight
+      else
+        return false # there are no existing reservations for tonight
+      end
```
Comment (kaidamasaki) [inline lib/date_range.rb:44]: You can just return the condition directly since it's already a boolean value:

```suggestion
      # true: there is an existing reservation for tonight
      # false: there are no existing reservations for tonight
      return range.include?(date)
```

File: lib/no_rooms_error.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+class NoAvailableRoomsError < StandardError
```
Comment (kaidamasaki) [inline lib/no_rooms_error.rb:1]: Custom error class!  :tada:

File: lib/reservation.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  class Reservation
+    attr_reader :date_range
+    attr_accessor :room_number
+    
+    # make room_number optional?
+    def initialize(date_range, room_number: nil, price_per_night: 200, total_cost: nil)
+      @date_range = date_range
+      @room_number = room_number
+      @price_per_night = price_per_night
+      @total_cost = total_cost
```
Comment (kaidamasaki) [inline lib/reservation.rb:13]: `total_cost` shouldn't be an argument to `Reservation` since you calculate it below so don't need to store it.  (And having it as an instance variable is misleading.)

File: lib/reservation_block.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    
+    def initialize(date_range, discount_rate, rooms:)
+      @date_range = date_range
+      
+      @rooms = rooms
+      validate_rooms
+      
+      @discount_rate = discount_rate
+      @total_block_cost = total_block_cost
+      @block_id = rand(100000.999999)
```
Comment (kaidamasaki) [inline lib/reservation_block.rb:16]: If you want a method that gives you an id that you can be confident won't repeat you can use [`SecureRandom#uuid`](https://ruby-doc.org/stdlib-1.9.3/libdoc/securerandom/rdoc/SecureRandom.html#method-c-uuid).

File: lib/reservation.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
+require_relative 'date_range'
+
+module Hotel
+  class Reservation
+    attr_reader :date_range
+    attr_accessor :room_number
```
Comment (kaidamasaki) [inline lib/reservation.rb:6]: I'm not clear on why `room_number` is an `attr_accessor`.

File: refactors.txt [hunk 10]
```diff
@@ -0,0 +1,12 @@
+Refactoring Hotel:
+- go through and be consistent about passing either a daterange or start/end dates
```
Comment (kaidamasaki) [inline refactors.txt:2]: 👍

File: refactors.txt [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+- update tests to reflect this
+
+- refactor so that FrontDesk holds a collection of @rooms, @reservations, & @blocks rather than rooms holding their reservations
+
+- make a choice about rooms; do you want them to be stored as objects or just numbers?
+
+- go through and choose a convention using .first or [0]
```
Comment (kaidamasaki) [inline refactors.txt:10]: I personally like `.first` and `.last` over `[0]` and `[-1]` since I think they're a little clearer but YMMV!

File: refactors.txt [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+- update tests to reflect this
+
+- refactor so that FrontDesk holds a collection of @rooms, @reservations, & @blocks rather than rooms holding their reservations
+
+- make a choice about rooms; do you want them to be stored as objects or just numbers?
+
+- go through and choose a convention using .first or [0] 
+
+- go through and try to be consistent about the order in which arguments are passed
```
Comment (kaidamasaki) [inline refactors.txt:12]: You can also just make everything keyword arguments.  😛

File: test/date_range_test.rb [hunk 13]
```diff
@@ -0,0 +1,137 @@
+require_relative 'test_helper'
+
+
+describe 'Date class' do
+  let (:date_range) {
+    Hotel::DateRange.new((Date.today), (Date.today + 3))
```
Comment (kaidamasaki) [inline test/date_range_test.rb:6]: Unnecessary parenthesis (it's already unambiguous):

```suggestion
    Hotel::DateRange.new(Date.today, Date.today + 3)
```

File: test/date_range_test.rb [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    expect(date_range).must_be_instance_of Hotel::DateRange
+  end
+  
+  it 'assigns check in correctly' do
+    expect(date_range.check_in).must_equal Date.today
+  end
+  
+  it 'assigns check out correctly' do
+    expect(date_range.check_out).must_equal Date.today + 3
+  end
```
Comment (kaidamasaki) [inline test/date_range_test.rb:19]: I might combine all of these into a single test but that's a matter of personal preference.  Just wanted to let you know that it's okay if you don't want to split things out so small.

File: test/date_range_test.rb [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+  
+  it "Calculates 1 night" do
+    range = Hotel::DateRange.new(Date.today, Date.today + 1)
+    expect(range.num_nights).must_equal 1
+  end
+  
+  it "Calculates many nights" do
+    range = Hotel::DateRange.new(Date.today, Date.today + 3)
+    expect(range.num_nights).must_equal 3
+  end
```
Comment (kaidamasaki) [inline test/date_range_test.rb:52]: I like describing these as "one" and "many".

**Comment by kaidamasaki** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | ✔️

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | ✔️
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | ✔️
creates a block of rooms | ✔️
reserves a room from a block | Did not have a method to directly reserve a room that's inside a block.

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections | ✔️
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you had a lot of different test cases for different types of nominal cases and your architecture was also quite clean. 

I do see some room for improvement around edge case tests (as you pointed out).  While 100% code coverage doesn't guarantee that you've hit every edge case generally speaking having uncovered code means you've missed some.  (You can of course have code that runs a line but doesn't check edgy behavior, like for an empty array as input or something.)

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise |
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Great job!  Here are some small ways you could clean up your project.
